File: server/notification/github/src/main/java/projektor/notification/github/GitHubCommentClient.java
package projektor.notification.github;

import org.kohsuke.github.*;

public class GitHubCommentClient {
    private final GitHubClientConfig clientConfig;
    private final JwtProvider jwtProvider;

    public GitHubCommentClient(GitHubClientConfig clientConfig, JwtProvider jwtProvider) {
        this.clientConfig = clientConfig;
        this.jwtProvider = jwtProvider;
    }

    public void addComment(String orgName, String repoName, int issueId, String commentText) throws Exception {
        String jwtToken = this.jwtProvider.createJWT(
                this.clientConfig.githubAppId,
                this.clientConfig.pemContents,
                this.clientConfig.ttlMillis
        );
        GitHub gitHub = new GitHubBuilder()
                .withJwtToken(jwtToken)
                .withEndpoint(this.clientConfig.gitHubApiUrl)
                .build();

        GHApp gitHubApp = gitHub.getApp();
        GHAppInstallation appInstallation = gitHubApp.getInstallationByRepository(orgName, repoName);
        GHAppInstallationToken appInstallationToken = appInstallation.createToken().create();

        GitHub githubAuthAsInst = new GitHubBuilder()
                .withAppInstallationToken(appInstallationToken.getToken())
                .withEndpoint(this.clientConfig.gitHubApiUrl)
                .build();

        GHRepository repository = githubAuthAsInst.getRepository(orgName + "/" + repoName);

        GHIssue ghIssue = repository.getIssue(issueId);
        ghIssue.comment(commentText);
    }
}

File: server/notification/github/src/main/kotlin/projektor/notification/github/GitHubCommentClient.kt
package projektor.notification.github

import org.kohsuke.github.*
import projektor.notification.github.auth.JwtProvider

class GitHubCommentClient(
    private val clientConfig: GitHubClientConfig,
    private val jwtProvider: JwtProvider
) {

    fun addComment(repository: GHRepository, issueId: Int, commentText: String) {
        val ghIssue = repository.getIssue(issueId)
        ghIssue.comment(commentText)
    }

    fun findCommentWithText(repository: GHRepository, issueId: Int, commentText: String): GHIssueComment? {
        val ghIssue = repository.getIssue(issueId)

        val comments = ghIssue.listComments().toList()

        return comments.find { it.body.contains(commentText) }
    }

    fun updateComment(comment: GHIssueComment, newText: String) {
        comment.update(newText)
    }

    fun findOpenPullRequestsForBranch(repository: GHRepository, branchName: String): Int? {
        val pullRequestsResults = repository
            .queryPullRequests()
            .state(GHIssueState.OPEN)
            .list()
            .toList()

        val pullRequestForBranch = pullRequestsResults
            .firstOrNull { it.head.ref.endsWith(branchName) }

        return pullRequestForBranch?.number
    }

    fun getRepository(orgName: String, repoName: String): GHRepository {
        val jwtToken = jwtProvider.createJWT()

        val gitHub = GitHubBuilder()
            .withJwtToken(jwtToken)
            .withEndpoint(this.clientConfig.gitHubApiUrl)
            .build()
        val gitHubApp = gitHub.app
        val appInstallation = gitHubApp.getInstallationByRepository(orgName, repoName)
        val appInstallationToken = appInstallation.createToken().create()
        val githubAuthAsInst = GitHubBuilder()
            .withAppInstallationToken(appInstallationToken.token)
            .withEndpoint(this.clientConfig.gitHubApiUrl)
            .build()
        val repository = githubAuthAsInst.getRepository("$orgName/$repoName")

        return repository
    }
}
